plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -1140,18 +1143,28 @@ public boolean downloadToFile(String urlString, Plugin plugin, @CheckForNull Fil
                     .setRetryHandler(new DefaultHttpRequestRetryHandler(maxRetries, true))
                     .build()) {
                 HttpClientContext context = HttpClientContext.create();
-                HttpHead httphead = new HttpHead(urlString);
-                try (CloseableHttpResponse ignored = httpclient.execute(httphead, context)) {
-                    HttpHost target = context.getTargetHost();
-                    List<URI> redirectLocations = context.getRedirectLocations();
-                    // Expected to be an absolute URI
-                    URI location = URIUtils.resolve(httphead.getURI(), target, redirectLocations);
-                    logVerbose(String.format("downloading %s from %s", plugin.getName(), location));
-                    FileUtils.copyURLToFile(location.toURL(), pluginFile);
-                } catch (URISyntaxException | IOException e) {
+                CredentialsProvider credentialsProvider = getCredentialsProvider();
+                if (credentialsProvider != null) {
+                    context.setCredentialsProvider(credentialsProvider);
+                }
+                HttpGet httpGet = new HttpGet(urlString);
+                try {
+                    httpclient.execute(httpGet, new FileDownloadResponseHandler(pluginFile), context);
+                } catch (IOException e) {
                     logVerbose(String.format("Unable to resolve plugin URL %s, or download plugin %s to file",
                             urlString, plugin.getName()));
                     success = false;
+                } finally {
+                    // get final URI (after all redirects)
+                    List<URI> locations = context.getRedirectLocations();
+                    if (locations != null) {
+                        String message = String.format("%s %s from %s", success ? "Downloaded" : "Tried downloading", plugin.getName(), locations.get(locations.size() - 1));
+                        if (success) {
+                            System.out.println(message);
+                        } else {
+                            logVerbose(message);
+                        }
```

Comment:
this appears to be around the wrong way?

Always log on failure, but verbose logging for the success case
```suggestion
                        if (success) {
                            logVerbose(message);
                        } else {
                            System.out.println(message);
                        }
```